protocols/request-response/src/cbor.rs

```diff
@@ -0,0 +1,303 @@
+// Copyright 2023 Protocol Labs
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{Config, ProtocolSupport, RequestId, ResponseChannel};
+use async_trait::async_trait;
+use futures::prelude::*;
+use futures::{AsyncRead, AsyncWrite};
+use libp2p_core::upgrade::{read_length_prefixed, write_length_prefixed};
+use libp2p_core::Multiaddr;
+use libp2p_identity::PeerId;
+use libp2p_swarm::StreamProtocol;
+use libp2p_swarm_derive::NetworkBehaviour;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{io, marker::PhantomData};
+
+/// Max request size in bytes
+const REQUEST_SIZE_MAXIMUM: usize = 1024 * 1024;
+/// Max response size in bytes
+const RESPONSE_SIZE_MAXIMUM: usize = 10 * 1024 * 1024;
+
+#[derive(Debug, Clone)]
+pub struct Codec<Req, Resp> {
+    phantom: PhantomData<(Req, Resp)>,
+}
+
+pub type OutEvent<Req, Resp> = crate::Event<Req, Resp>;
+
+#[derive(NetworkBehaviour)]
+#[behaviour(
+    to_swarm = "OutEvent<Req, Resp>",
+    prelude = "libp2p_swarm::derive_prelude"
+)]
+pub struct Behaviour<Req, Resp>
+where
+    Req: Send + Clone + Serialize + DeserializeOwned + 'static,
+    Resp: Send + Clone + Serialize + DeserializeOwned + 'static,
+{
+    inner: crate::Behaviour<Codec<Req, Resp>>,
+}
+
+fn convert_error(err: serde_cbor::Error) -> io::Error {
+    if err.is_syntax() || err.is_data() {
+        io::Error::new(io::ErrorKind::InvalidData, err)
+    } else if err.is_eof() {
+        io::Error::new(io::ErrorKind::UnexpectedEof, err)
+    } else {
+        io::Error::new(io::ErrorKind::Other, err)
+    }
```

Comment:
```suggestion
    if err.is_syntax() || err.is_data() {
        return io::Error::new(io::ErrorKind::InvalidData, err)
    }
    
    if err.is_eof() {
        return io::Error::new(io::ErrorKind::UnexpectedEof, err)
    }
    
    io::Error::new(io::ErrorKind::Other, err)
```

Prefer not using `else`.